source/fe/mapping.cc

```diff
@@ -36,6 +36,21 @@ Mapping<dim, spacedim>::get_vertices(
 
 
 
+template <int dim, int spacedim>
+Point<spacedim>
+Mapping<dim, spacedim>::get_center(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell) const
+{
+  const auto vertices = get_vertices(cell);
+  AssertDimension(vertices.size(), GeometryInfo<dim>::vertices_per_cell);
+  Point<spacedim> center;
+  for (const auto &v : vertices)
+    center += v;
+  return center / GeometryInfo<dim>::vertices_per_cell;
```

Comment:
Shouldn't you map the average instead of averaging the mapped vertices?

Reply:
You mean like Mapping::transform_unit_to_real(Point<dim>(.5, .5));? 

Yes, this makes more sense, now that I think about this...

Reply:
I think that the original implementation was right. If the mapping is non-linear, then the centre of the unit cell need not be mapped to the centre of the deformed cell. And you're wanting to return the centre of the deformed cell, right?

Reply:
@jppelteret , yes, that's what I want. I mean, it all really depends on what we define as the cell center. The test seems to think you are right, but I fear that the test simply uncovered a bug in MappingQEulerian. Think of this case:

![image](https://user-images.githubusercontent.com/2919756/56343670-a82fe300-61bb-11e9-9198-1b0c0e5f87c8.png)

The dot below is the average of the deformed vertices, while the dot above should be the mapping of the center in the reference cell. I tend to think we should return the dot above, and not the dot below, but currently this is not working. 

What do I do? Do I revert back to what I had before? Do we keep it as is now, and we try to figure out what's wrong with MappingQEulerian?

Reply:
Hmm... Your extreme example certainly highlights the issue here. I guess that neither approaches are correct then, as I also would have expected the above dot to be the one returned. Maybe the solution is non-trivial, and we need to solve for the centre point in the convected coordinate system?

Reply:
> the centre point in the convected coordinate system
Isn't this exactly what 
~~~
Mapping::transform_unit_to_real_cell(cell, Point<2>(.5, .5));
~~~
should return in the case of `MappingQEulerian`?

Reply:
Maybe I'm misunderstanding everything completely, but here's an example that (I think) [demonstrates my doubt](https://gist.github.com/jppelteret/1f1f3f800a287b45c9fe89143b23bd43). The tria has one cell with vertices coinciding with the unit cell. A quadratic FE space is used and a quadratic displacement map is applied. The unit cell is now stretched by a factor of 2 in the x direction, but in a non-affine manner.

Unless I'm mistaken in my interpretation of how things are supposed to work, `Mapping::transform_unit_to_real_cell(cell, Point<2>(.5, .5));` does not return the centre of the deformed cell (which should be `(1,0.5)`) but rather `(0.75,0.5)`. The point that maps to `(1,0.5)` in the deformed configuration is something else.

Am I mistaken in my thinking here? I wouldn't be surprised, given my lack of sleep...

Reply:
I think `(0.75,0.5)` is totally correct. I would have expected that as midpoint.

Reply:
On the other hand, I am not sure that I understand the test output.

Reply:
I have reverted to the old implementation. This at least works fine for standard cases, and it is more correct w.r.t. what we had. 

The picture of the transformation is the following (red dot is (.5, 0)):

![image](https://user-images.githubusercontent.com/2919756/56415285-8c950d00-628d-11e9-93a2-53e81b899931.png)

Reply:
By the way, I think that the output was correct also for the other implementation. Now I'm convinced that it does not make much sense (i.e., the point (0,0), which is the original cell center, is a fixed point of the transformation in the test, and it is clearly *not* the center of the deformed cell). As is documented, this function is now correct.

Reply:
I think it all depends on how you define what the center is, but returning `Mapping::transform_unit_to_real_cell(cell, Point<2>(.5, .5));` would seem like the right choice to me.